### PR TITLE
Add Auth MonitorConnections chain elements

### PR DESCRIPTION
**Pull request by anastasia-malysheva**
## Description

Add Authorization Chain element for MonitorConnection {Client/Server}.

- [x] Add next MonitorConnections elements
- [x] Add auth MonitorConnections elements
- [x] Add usage of new MonitorConnections elements to the existing ones
- [x] Add Policies for authorizations

## Issue link
https://github.com/networkservicemesh/sdk/issues/46

## How Has This Been Tested?

- [x] Added unit testing to cover
- [x] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/chains/nsmgr/server.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ import (
 	registryadapter "github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
 	"github.com/networkservicemesh/sdk/pkg/registry/core/chain"
 	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+	authMonitor "github.com/networkservicemesh/sdk/pkg/tools/monitor/authorize"
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server.go:61]: Please use lowercase package naming to follow effective go conventions.

File: pkg/networkservice/chains/nsmgr/server.go [hunk 2]
```diff
@@ -75,6 +76,7 @@ type nsmgrServer struct {
 
 type serverOptions struct {
 	authorizeServer      networkservice.NetworkServiceServer
+	authMonitorOptions   []authMonitor.Option
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server.go:79]: Please do the same as we do for auth networkservice.NetworkServiceServer

**Review by denis-tingaikin — COMMENTED** [review]
Started looking into this PR. Added a few comments, I'll look deeper a bit later.

File: pkg/networkservice/common/heal/eventloop.go [hunk 3]
```diff
@@ -63,7 +64,8 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
 		},
 	}
 
-	client, err := networkservice.NewMonitorConnectionClient(cc).MonitorConnections(eventLoopCtx, selector)
+	mClient := next.NewMonitorConnectionClient(networkservice.NewMonitorConnectionClient(cc))
+	client, err := mClient.MonitorConnections(eventLoopCtx, selector)
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/heal/eventloop.go]: removed

File: pkg/networkservice/common/monitor/eventloop.go [hunk 4]
```diff
@@ -51,7 +53,8 @@ func newEventLoop(ctx context.Context, ec EventConsumer, cc grpc.ClientConnInter
 		},
 	}
 
-	client, err := networkservice.NewMonitorConnectionClient(cc).MonitorConnections(eventLoopCtx, selector)
+	client, err := next.NewMonitorConnectionClient(
+		networkservice.NewMonitorConnectionClient(cc)).MonitorConnections(eventLoopCtx, selector)
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/monitor/eventloop.go]: removed

File: pkg/networkservice/core/eventchannel/monitor_client.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2020 Cisco and/or its affiliates.
+// Copyright (c) 2020-2022 Cisco and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/core/eventchannel/monitor_client.go]: removed

File: pkg/networkservice/common/authorize/server_test.go [hunk 6]
```diff
@@ -103,11 +104,11 @@ func TestAuthzEndpoint(t *testing.T) {
 			denied:  true,
 		},
 	}
-
+	spiffeIDConnectionMap := spire.SpiffeIDConnectionMap{}
 	for i := range suits {
 		s := suits[i]
 		t.Run(s.name, func(t *testing.T) {
-			srv := authorize.NewServer(authorize.WithPolicies(s.policy))
+			srv := authorize.NewServer(authorize.WithSpiffeIDConnectionMap(&spiffeIDConnectionMap), authorize.WithPolicies(s.policy))
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/authorize/server_test.go]: removed

File: pkg/networkservice/common/authorize/server.go [hunk 7]
```diff
@@ -21,48 +21,63 @@ package authorize
...
+		PathSegments: conn.GetPath().GetPathSegments()[:index+1],
 	}
 	if _, ok := peer.FromContext(ctx); ok {
 		if err := a.policies.check(ctx, leftSide); err != nil {
 			return nil, err
 		}
 	}
+	if spiffeID, err := getSpiffeID(ctx); err == nil {
+		ids, _ := a.spiffeIDConnectionMap.Load(spiffeID)
+		a.spiffeIDConnectionMap.LoadOrStore(spiffeID, append(ids, conn.GetId()))
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/authorize/server.go]: I've added nested map. Please take a look.

File: pkg/networkservice/chains/nsmgr/server.go (hunk 2 above)
Comment (anastasia-malysheva) [reply pkg/networkservice/chains/nsmgr/server.go]: fixed

File: pkg/networkservice/common/monitor/server_test.go [hunk 8]
```diff
@@ -32,10 +37,39 @@ import (
...
+aMp+T747AZGjOEfwHb9/w+7m
+-----END CERTIFICATE-----
+`
+)
+
 func TestMonitorServer(t *testing.T) {
 	t.Cleanup(func() { goleak.VerifyNone(t) })
 
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	// Put peer Certificate to context
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/monitor/server_test.go]: removed

File: pkg/tools/monitor/authorize/common.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+type policiesList []Policy
+
+func (l *policiesList) check(ctx context.Context, srv MonitorOpaInput) error {
+	if l == nil {
+		return nil
+	}
+	for _, policy := range *l {
+		if policy == nil {
```
Comment (denis-tingaikin) [inline pkg/tools/monitor/authorize/common.go:36]: When have we a list with a nil policy?

File: pkg/tools/monitor/next/client.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2022 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/tools/monitor/next/client.go:1]: Do we need a next client for this PR?

File: pkg/tools/monitor/next/client_test.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2022 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/tools/monitor/next/client_test.go:1]: Do we need a next client for this PR?

File: pkg/tools/opa/opainput.go [hunk 12]
```diff
@@ -55,7 +54,8 @@ func pemEncodingX509Cert(cert *x509.Certificate) string {
 	return string(certpem)
 }
 
-func parseX509Cert(authInfo credentials.AuthInfo) *x509.Certificate {
+// ParseX509Cert - parse authinfo to get peer sertificate
```
Comment (denis-tingaikin) [inline pkg/tools/opa/opainput.go:57]:
```suggestion
// ParseX509Cert parses x509 certificate from the passed  credentials.AuthInfo
```

File: pkg/tools/opa/policies.go [hunk 13]
```diff
@@ -77,3 +80,11 @@ func WithTokensExpiredPolicy() *AuthorizationPolicy {
...
 	}
 }
+
+func WithServiceOwnConnectionPolicy() *AuthorizationPolicy {
+	return &AuthorizationPolicy{
+		policySource: tokensServiceConnectionPolicySource,
+		query:        "service_connection",
+		checker:      True("service_connection"),
+	}
+}
```
Comment (denis-tingaikin) [inline pkg/tools/opa/policies.go:90]: I think we need consider a better name to clarify a fact that the policy should be used only and only with monitors.

File: pkg/tools/opa/service_connection_policy_test.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/monitor/authorize"
+	"github.com/networkservicemesh/sdk/pkg/tools/opa"
+)
+
+func TestWithServiceConnectionPolicy(t *testing.T) {
```
Comment (denis-tingaikin) [inline pkg/tools/opa/service_connection_policy_test.go:29]: Could we please add a test for a negatvie scenario?

File: pkg/tools/spire/gen_conn_map.go [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+package spire
+
+import (
+	"sync"
+)
+
+//go:generate go-syncmap -output connection_map.gen.go -type ConnectionMap<string,bool>
+
+// ConnectionMap - sync.Map with key == string and value == bool
+type ConnectionMap sync.Map
```
Comment (denis-tingaikin) [inline pkg/tools/spire/gen_conn_map.go:26]: This is not looking as a connection map/

I think it is a simple string set.

**Review by denis-tingaikin — COMMENTED** [review]
Conceptually looks correct

File: pkg/tools/monitor/authorize/common.go (hunk 9 above)
Comment (anastasia-malysheva) [reply pkg/tools/monitor/authorize/common.go:36]: option Any put nil as a policy

File: pkg/networkservice/common/monitor/monitor_connection_server.go [hunk 16]
```diff
@@ -22,24 +22,31 @@ import (
 	"github.com/edwarnicke/serialize"
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/monitor/monitor_connection_server.go:22]: In current solution we actually dont need changes in `network service/monitor` pkg.

Comment (denis-tingaikin) [reply pkg/networkservice/common/monitor/monitor_connection_server.go:22]: Please revet all changes in monitor pkg to avoid new issues.

Comment (denis-tingaikin) [inline pkg/networkservice/common/monitor/monitor_connection_server.go:22]: In current solution we actually dont need changes in network service/monitor pkg. Please revert pkg/networkservice/monitor changes.

File: pkg/networkservice/common/monitor/server.go [hunk 17]
```diff
@@ -23,6 +23,7 @@ package monitor
 import (
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/monitor/server.go:23]: In current solution we actually dont need changes in network service/monitor pkg. Please revert pkg/networkservice/monitor changes.

File: pkg/tools/monitor/authorize/common.go [hunk 18]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2022 Cisco and/or its affiliates.
```
Comment (denis-tingaikin) [inline pkg/tools/monitor/authorize/common.go:1]: Should we rename pkg tools/monitor into tools/monitorconnection?

File: pkg/networkservice/common/authorize/server.go [hunk 19]
```diff
@@ -72,10 +97,37 @@ func (a *authorizeServer) Close(ctx context.Context, conn *networkservice.Connec
...
+	cert = opa.ParseX509Cert(p.AuthInfo)
+	if cert != nil {
+		spiffeID, err := x509svid.IDFromCert(cert)
+		if err == nil {
+			return spiffeID.String(), nil
+		}
+		return "", errors.New("fail to get Spiffe ID from certificate")
+	}
+	return "", errors.New("fail to get certificate from peer")
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/authorize/server.go:133]: I think we should move this into tools/spire

File: pkg/networkservice/common/authorize/server.go [hunk 20]
```diff
@@ -72,10 +97,37 @@ func (a *authorizeServer) Close(ctx context.Context, conn *networkservice.Connec
...
+	}
 	if _, ok := peer.FromContext(ctx); ok {
 		if err := a.policies.check(ctx, leftSide); err != nil {
 			return nil, err
 		}
 	}
 	return next.Server(ctx).Close(ctx, conn)
 }
+
+func getSpiffeID(ctx context.Context) (string, error) {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/authorize/server.go:118]:
```suggestion
func SpiffeIDFromContext(ctx context.Context) (string, error) {
```

File: pkg/tools/opa/opainput.go [hunk 21]
```diff
@@ -55,7 +54,8 @@ func pemEncodingX509Cert(cert *x509.Certificate) string {
 	return string(certpem)
 }
 
-func parseX509Cert(authInfo credentials.AuthInfo) *x509.Certificate {
+// ParseX509Cert - parses x509 certificate from the passed  credentials.AuthInfo
```
Comment (NikitaSkrynnik) [inline pkg/tools/opa/opainput.go:57]:
```suggestion
// ParseX509Cert - parses x509 certificate from the passed credentials.AuthInfo
```

File: pkg/tools/spire/gen_conn_map.go [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+package spire
+
+import (
+	"sync"
+)
+
+//go:generate go-syncmap -output connection_id_set.gen.go -type ConnectionIDSet<string,bool>
+
+// ConnectionIDSet - sync.Map with key == string and value == bool
+type ConnectionIDSet sync.Map
```
Comment (denis-tingaikin) [inline pkg/tools/spire/gen_conn_map.go:26]: Why do we need to store this in tools/spire?

Can we move this into tools/stringset?

Also please use zero value type `struct{}`.

**Review by NikitaSkrynnik — APPROVED** [review]

File: pkg/tools/opa/policies.go (hunk 13 above)
Comment (anastasia-malysheva) [reply pkg/tools/opa/policies.go:90]: WithMonitorConnectionServerPolicy

File: pkg/networkservice/common/monitor/server.go [hunk 23]
```diff
@@ -50,10 +54,17 @@ type monitorServer struct {
...
 func NewServer(chainCtx context.Context, monitorServerPtr *networkservice.MonitorConnectionServer) networkservice.NetworkServiceServer {
-	*monitorServerPtr = newMonitorConnectionServer(chainCtx)
+	filters := make(map[string]*monitorFilter)
+	executor := serialize.Executor{}
+	connections := make(map[string]*networkservice.Connection)
+
+	*monitorServerPtr = newMonitorConnectionServer(chainCtx, &executor, filters, connections)
 	return &monitorServer{
 		chainCtx:                chainCtx,
 		MonitorConnectionServer: *monitorServerPtr,
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/monitor/server.go]: fixed

File: pkg/networkservice/common/authorize/server.go [hunk 24]
```diff
@@ -72,10 +97,37 @@ func (a *authorizeServer) Close(ctx context.Context, conn *networkservice.Connec
...
 	}
+	if spiffeID, err := getSpiffeID(ctx); err == nil {
+		connID := conn.GetPath().GetPathSegments()[index-1].GetId()
+		ids, ok := a.spiffeIDConnectionMap.Load(spiffeID)
+		if ok {
+			if present, ok := ids.Load(connID); present && ok {
+				ids.Delete(connID)
+			}
+		}
+		a.spiffeIDConnectionMap.Store(spiffeID, ids)
```
[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply pkg/networkservice/common/authorize/server.go]: fixed

File: pkg/tools/spire/gen_conn_map.go (hunk 22 above)
Comment (anastasia-malysheva) [reply pkg/tools/spire/gen_conn_map.go:26]: fixed

**Review by glazychev-art — APPROVED** [review]

**Review by denis-tingaikin — APPROVED** [review]

**Comment by denis-tingaikin** [comment]
This is a bit reworked and merged in https://github.com/networkservicemesh/sdk/pull/1333

**Pull request closed without merging**